src/Ormolu/Printer/Meat/Declaration/Value.hs

```diff
@@ -249,10 +249,17 @@ p_match' placer render style isInfix strictness m_pats GRHSs {..} = do
       placement =
         case endOfPats of
           Nothing -> blockPlacement placer grhssGRHSs
-          Just spn ->
-            if onTheSameLine spn grhssSpan
-              then blockPlacement placer grhssGRHSs
-              else Normal
+          Just spn -> case style of
+            Function _ | hasGuards && any guardNeedsLineBreak grhssGRHSs -> Normal
```

Comment:
I am not sure if this was discussed, but is it intended that this change is only applied for the `Function` style? Example: Even with this PR, the input
```haskell
\case x | x <-
          foo -> case x of
         5 -> True
         _ -> False
```
is formatted to
```haskell
\case
  x | x <-
        foo -> case x of
    5 -> True
    _ -> False
```
(note that there is no linebreak before the `|`). Simply ignoring `style` passes all tests locally, and the output would be
```haskell
\case
  x
    | x <-
        foo -> case x of
      5 -> True
      _ -> False
```
which now is consistent with the output for regular function match groups.

Reply:
That's a good point indeed, and we didn't discussed it before. However, I don't have the authority to decide whether we should extend it to all styles or not :)
But IMO, it's better to extend it!